### ISSUE TITLE
🪣 Proposal <> Thread linking should not require Chain Events

### ISSUE THREAD

**Issue by jnaviask** [hicommonwealth/commonwealth#3357]
## Description
Currently we support linking proposals to threads. This link is bidirectional, i.e. you can see a link to a proposal page within the same community from a thread page, and vice versa.

## Outstanding Tickets
- [x] #3539 
- [x] #3540 
- [x] #3541 
- [x] #3645
- [x] #3646
- [x] #3674

## Technical Requirements

The current design of the system is, in order to link between a Thread and an on-chain Proposal, the on-chain Proposal must be present in our database as a ChainEntityMeta row (see `updateThreadLinkedChainEntities.ts` route), so that the ChainEntityMeta in question's `thread_id` column points to the `Thread`'s id.

We want to move Proposal Linking to a system closer to the way we link Threads with Snapshot Proposals (see `updateThreadLinkedSnapshotProposal` route), but without deprecating any functionality, and ideally without introducing too many more relational links in the database.

Linking has the following requirements (briefly):
1. Linking user must see a list of possible proposals they can link to a given thread (currently part of `ChainEntitiesSelector.tsx`), and must be able to select >= 1 of them.
2. Selected proposals must appear as links in a box on the thread's page.
3. Thread must view as a link on the corresponding linked proposal's page.

Following from the product requirements is a list of engineering requirements:
1. Linking modal must be able to access a list of proposals currently extant within a community.
2. Thread pages must be able to access a list of links to proposal pages based on selection, if any.
3. Proposals pages must be able to access a link to a thread page, if it exists.

In theory, we can satisfy these engineering requirements with the following:

1. If a community can load proposals, then these proposals are accessible via a store on the front-end, and at minimum _have a unique identifier_ (from which we can produce links `/proposal/<identifier>`). These proposals should be loaded within the modal as if we were loading the proposals page (i.e. via a more abstract layer), rather than accessing the underlying store directly and assuming chain events. _This part still TBD in terms of how its implementation should look._
2. Thread pages should be able to ask the server for a list of links based on community and thread id, which may be tagged e.g. `[{ type: 'snapshot", identifier: '0xabcdef12345' }, { type: 'proposal', identifier: '5' }, { type: 'thread', identifier: 1234 }, { type: 'web', url: 'www.mywebsite.com' }]`, and then are rendered in determined areas on the thread page (_need product / design feedback here_).
3. Proposal pages should be able to ask the server for a list of links based on community and proposal identifier, which may be in a similar form as on thread pages (so we can reuse the component).

The solution of storing links rather than entity references abstracts away the requirement for entity existence in our database, and makes the linking functionality more powerful.

cc @zakhap to review the above for product considerations.

## Acceptance Criteria
- Requirements remain met.
- Linking is possible in communities where we do not have on-chain proposals stored in our database.

**Comment by CowMuon** [comment]
This needs to be in scope for discussion. We won't start until next sprint (sprint #3 starts 4/17) but just to get it on our radar. 

Proposal linking is literally one of the most important things we do. It's at the nucleus of probably our most important value prop. By way of saying we need to follow Michael B's "first day advice" (see glossary under D.F.I.U., loosely translated := **Important to get right**.

Teed Up without Assignee, which we usually don't do, but in this case would like to get the ball rolling.

**Comment by jnaviask** [comment]
In Cosmos communities, we load proposals directly from chain in `controllers/chain/cosmos/governance.ts`. Sputnik proposals on NEAR work similarly.

**Comment by ianrowan** [comment]
A thought/idea here
- convert `snapshot_proposal` column to `proposal`
- store a 'type' prefixed encoded(snapshot is already encoded) proposal number/identifier
- `type` = first `n` bytes of hex string(based on how many types we expect to have)
- So a `proposal` id looks like `0x[type_byte]+uniqueId`

- Then to retrieve a linked proposal we have enums/processors for each start byte(ie cosmos byte = decode 32 byte hex to int for byte 1 -> 32(0 indexed)  and snapshot just returns byte 1+ as a snapshot id) and a response from server for thread links can be returned like `{ 'cosmosProposol': 479}` or `{'snapshotProposal': "0x12345"}` 
- upon linking FE can upload a proposal Id(non-encoded) + proposal type and BE can handle the storage 

Might be reaching for some EVM/Smart contract data structures but they are moderately efficient and could allow us to avoid adding any columns or tables to the DB.

TLDR: We store our own identifier in a column on the thread that can be decoded to the proposal type 

@jnaviask @CowMuon

**Comment by jnaviask** [comment]
Does this support multiple links?

Also I'm a little wary of using byte-based prefixing to save on data, when we should be optimizing our designs for maintainability, imo (until we reach the scale where we need the data-saving optimizations).

**Comment by ianrowan** [comment]
> Does this support multiple links?
> 
> Also I'm a little wary of using byte-based prefixing to save on data, when we should be optimizing our designs for maintainability, imo (until we reach the scale where we need the data-saving optimizations).

@jnaviask could support multiple links if we wanted given a type could define how many bytes its encoded into, then just string as many type/data combos in the string to represent multiple types/keys

but I see the point of maintainability. From that perspective would we want to consider storing json in the Thread table or something more formal like a junction table `Thread_Links` That can maintains records of thread id foreign key -> proposalId relationships?

**Comment by jnaviask** [comment]
I'm happy with JSON in the thread table. We want to avoid too much relationality in our db schema, as it's hard to port over to protocol long-term.

**Comment by dillchen** [comment]
From the product perspective, and perhaps not on this PR, is that `Links` should be a type of content, and live as their own DBModel.

Links are bidirectional, and right now live on the thread model. This feels weird for example on `View_Snapshot` we

`fetchThreadIdsForSnapshot(args: { snapshot: string }) {` to present the ThreadIDs and titles. 

In the future, Snapshots may be directly linked to Proposals without a Thread. Images (NFTs) may be linked to other things without a Thread.

**Comment by ianrowan** [comment]
First prototype here: https://github.com/hicommonwealth/commonwealth/commit/8e802bb9a3c4c3715df3463e8dc5029f671adb0d
creates this while not breaking functionality:

<img width="730" alt="image" src="https://user-images.githubusercontent.com/36428666/233123637-b3e24cdf-d979-49a2-8afb-061e5578c8d3.png">

idea would then be to add an equivalent `updateThreadLinkedSnapshotProposal ` for other types of proposals + valid read methods to pull that json blob to FE(which will know how to visualize a proposal based on source/id objects)

Thoughts? @jnaviask

**Comment by jnaviask** [comment]
Great breakdown, thanks @ianrowan. Would like to discuss with you and Forest before we move forward on picking one or the other. For the time being, can we add a small helper function / abstraction layer, so we can start building and decide on the implementation detail later?

**Comment by ianrowan** [comment]
@jnaviask Thanks! Sounds good. So far I've built the JSONB approach(as a prototype) into the snapshot, chainEntity, and a new generic proposal route here + a helper on the client thread controller https://github.com/hicommonwealth/commonwealth/tree/ian.proposal_linking. We would only need to modify the couple lines of code that write to the JSON column to write to the New table if we decide on that approach or other. Is that what you had in mind for abstraction here?

The FE is a bit more complicated and probably worth talking through w/ everyone(maybe including product). Would it make sense to split the data management/BE and the FE UI tasks into two issues so they can be worked on concurrently? (ie I continue building routes/helpers for data writing and retrieval and somebody else concurrently redesigns the modal and bidirectional links on the proposal pages) Wasnt sure what the plan was on this point

**Comment by ianrowan** [comment]
Update on the JSONB approach: attempting to prototype a bidirectional fetch of proposal -> many threads and JSON array parsing given the proposed Data structure. Postgres and sequelize don't have great built in ability to parse for matches within a list of objects in these columns. Could be a dealbreaker for storing arrays with the requirement to lookup a row where a single member of the array matches, etc

**Comment by jnaviask** [comment]
@ianrowan This is pretty close to what I had in mind! There's one other requirement here, which is that we also support linking threads _to each other_, so we will need to include Threads in the list of possible sources. I also wonder if there is a way to circumvent the ChainEvents dependency here, so we can refer to a proposal generically, and then IF it's available via ChainEntity, we're happy -- then we don't need to distinguish at the top level between links to ChainEntities and links to CosmosProposals.

I would also like to see the interface generalized a bit (+ using the TypedRequest/TypedResponse so we can see the interface directly), as it feels weird to have so many specific routes, when we can nail down the functionality as "I am a source of type [thread|proposal|etc], get me the [threads|proposals|etc]".

Then we end up with a CRUD layout, which we should use directly (e.g. as methods on `/linking` or whatever name feels best), where we can create/modify/delete links, and also get links for any source type.

**Comment by ianrowan** [comment]
Some stubs here + failing unit(really integration) tests. 
https://github.com/hicommonwealth/commonwealth/compare/master...ian.proposal_link_fe_stubs

I essentially just wrote the methods FE will use to interact with the server + the calls to the server but omitted any intermediate logic for error handling, return, etc. I did include suggested return types for each methods, also so that the mock unit tests could be written. 

Also marked up the FE controller with some comments as to what should/will be replace by this new logic

CC: @jnaviask @CowMuon

**Comment by jnaviask** [comment]
@ianrowan awesome, feel free to put up a PR with the stubs that closes #3539!

**Comment by ianrowan** [comment]
We'll use this as a feature branch(Draft PR linked too) https://github.com/hicommonwealth/commonwealth/tree/feature.thread_proposal_links

**Comment by jnaviask** [comment]
@ianrowan @masvelio Just wanted to give a quick extension of this spec, shouldn't take more than a few minutes. Main problem with above is that it doesn't support Substrate, where we can have many different types of proposals. I suggest the following solution:
- For @ianrowan: Add the `customSlug` field as optional on the `Link` object. This is populated by e.g. `democracyproposal` or `referendum` for Substrate proposals, and otherwise left blank. This field will need to be populated to the migration (should be the prefix on the proposal identifier).
- For @masvelio: If the `customSlug` field is present on the `Link` object on a Link of type `proposal`, generate a slug that looks like `/proposal/${customSlug}/${identifier}` instead of just `/proposal/${identifier}`.

Should be a straightforward update on both sides. Let me know if you need additional information or context to get this out the door.

**Comment by jnaviask** [comment]
Given that ChainEntityMeta does not store any information about the proposal that would let us derive the slug, we should not proceed with this change as specified. No need to add the `customSlug` field. We may need to drop Substrate proposal linking support.

**Comment by ianrowan** [comment]
As discussed in todays CE office hrs we want to support 2 new flows
- Link titles
    - BE: Add a title field to links 
    - FE: Use title field for UI purposes + populate via FE data or user input
- Redirects and updates for link building of non-prebuilt links.
    - There are 2 scenarios identified that require additional steps to produce a hyperlink for a link, snapshots missing space and ChainEntityMeta links. For each of these we need to build redirects to find the snapshot space, and proposal id respectively. After the first redirect the link should be updated so we dont need to build it more than once
    - BE: Add ability to update links 
    - FE: Build Redirect + inference of what source/identifier/chain combos to be redirected(ie a proposal from a CE chain that isnt a hyperlink yet) + link building logic + upadate

CC: @jnaviask @CowMuon
